packages/controller/main.js

```diff
@@ -2230,31 +2249,36 @@ async function processMessage(msg) {
                 }
                 logger.info(`${hostLogPrefix} ${tools.appName} ${args.slice(1).join(' ')}`);
 
-                const child = spawn('node', args, { windowsHide: true });
-                if (child.stdout) {
-                    child.stdout.on('data', data => {
-                        data = data.toString().replace(/\n/g, '');
-                        logger.info(hostLogPrefix + ' ' + tools.appName + ' ' + data);
-                        msg.from && sendTo(msg.from, 'cmdStdout', { id: msg.message.id, data: data });
-                    });
-                }
+                try {
+                    const child = spawn('node', args, { windowsHide: true });
+                    if (child.stdout) {
+                        child.stdout.on('data', data => {
+                            data = data.toString().replace(/\n/g, '');
+                            logger.info(`${hostLogPrefix} ${tools.appName} ${data}`);
+                            msg.from && sendTo(msg.from, 'cmdStdout', { id: msg.message.id, data: data });
+                        });
+                    }
 
-                if (child.stderr) {
-                    child.stderr.on('data', data => {
-                        data = data.toString().replace(/\n/g, '');
-                        logger.error(`${hostLogPrefix} ${tools.appName} ${data}`);
-                        msg.from && sendTo(msg.from, 'cmdStderr', { id: msg.message.id, data: data });
+                    if (child.stderr) {
+                        child.stderr.on('data', data => {
+                            data = data.toString().replace(/\n/g, '');
+                            logger.error(`${hostLogPrefix} ${tools.appName} ${data}`);
+                            msg.from && sendTo(msg.from, 'cmdStderr', { id: msg.message.id, data: data });
+                        });
+                    }
+
+                    child.on('exit', exitCode => {
+                        logger.info(`${hostLogPrefix} ${tools.appName} exit ${exitCode}`);
+                        if (msg.from) {
+                            sendTo(msg.from, 'cmdExit', { id: msg.message.id, data: exitCode });
+                            // Sometimes finished command is lost, recent it
+                            setTimeout(() => sendTo(msg.from, 'cmdExit', { id: msg.message.id, data: exitCode }), 1000);
+                        }
                     });
+                } catch (e) {
+                    logger.error(`${hostLogPrefix} ${tools.appName} ${e.message}`);
+                    msg.from && sendTo(msg.from, 'cmdStderr', { id: msg.message.id, data: e.message });
```

Comment:
^^

Reply:
??

Reply:
Nit: true && fun()

i know it is on the legacy code above too but we could change if we add new lines ;-)

Reply:
ok, next time